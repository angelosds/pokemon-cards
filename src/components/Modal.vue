<template>
  <div class="modal">
    <section class="modal__container">
      <h2 class="modal__title">{{ title }}</h2>
      <div class="modal__content">
        <slot />
      </div>

      <div class="modal__footer">
        <primary-button
          class="modal__button modal__button--close"
          type="button"
          @on-click="onClose"
          >{{ $t("close") }}</primary-button
        >
      </div>
    </section>
  </div>
</template>

<script lang="ts">
import { Component, Emit, Prop, Vue } from "vue-property-decorator";

import PrimaryButton from "@/components/PrimaryButton.vue";

@Component({
  components: {
    PrimaryButton
  }
})
export default class Modal extends Vue {
  @Prop() private readonly title: string;

  @Emit()
  onClose(event: Event) {
    return event;
  }
}
</script>

<style lang="scss" scoped>
.modal {
  align-items: center;
  background-color: $modal-background-color;
  display: flex;
  height: 100vh;
  justify-content: center;
  left: 0;
  padding: 24px;
  position: fixed;
  top: 0;
  width: 100vw;
  z-index: 10;

  &__container {
    background-color: $background-color;
    border-radius: 16px;
    max-width: 460px;
    width: 100%;
  }

  &__content {
    padding: 24px;
  }

  &__footer {
    border-top: 1px solid $secondary-color;
    display: flex;
    justify-content: flex-end;
    padding: 16px 24px;
  }

  &__title {
    border-bottom: 1px solid $secondary-color;
    font-size: 20px;
    font-weight: 700;
    margin: 0;
    padding: 16px 24px;
  }
}
</style>
