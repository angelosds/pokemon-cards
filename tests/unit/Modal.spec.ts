import { mount } from "@vue/test-utils";

import Modal from "@/components/Modal.vue";

describe("Modal.vue", () => {
  const build = (propsData?: object, slots?: object) => {
    const wrapper = mount(Modal, {
      propsData: { title: "Modal", ...propsData },
      slots: { ...slots },
      mocks: {
        $t: (text: string) => text
      }
    });

    const button = () => wrapper.find(".modal__button");
    const content = () => wrapper.find(".modal__content");
    const title = () => wrapper.find(".modal__title");

    return { button, content, title, wrapper };
  };

  it("renders the component", () => {
    const { wrapper } = build();

    expect(wrapper).toMatchSnapshot();
  });

  it("renders a title", () => {
    const titleText = "Attacks summary";
    const { title } = build({ title: titleText });

    expect(title().text()).toMatch(titleText);
  });

  it("renders a content in a default slot", () => {
    const contentText = "Attack list";
    const { content } = build(undefined, { default: contentText });

    expect(content().html()).toMatch(contentText);
  });

  it("emits on-close event", () => {
    const { button, wrapper } = build();

    button().trigger("click");

    expect(wrapper.emitted("on-close")).toBeTruthy();
  });
});
