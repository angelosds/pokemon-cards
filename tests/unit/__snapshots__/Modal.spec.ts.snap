// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.vue renders the component 1`] = `
<section class="modal"><span class="modal__title">Modal</span>
  <div class="modal__content"></div>
  <div class="modal__footer"><button type="button" class="primary-button modal__button">close</button></div>
</section>
`;
